source/http_message.c

```diff
@@ -0,0 +1,235 @@
+        }
+
+        path = PyObject_GetAttrString(message_self, "path");
+        struct aws_byte_cursor path_cur = aws_byte_cursor_from_pystring(path);
+        if (!path_cur.ptr) {
+            PyErr_SetString(PyExc_TypeError, "HttpRequest.path is invalid");
+            goto done;
+        }
+        if (aws_http_message_set_request_path(message->native, path_cur)) {
+            PyErr_SetAwsLastError();
+            goto done;
+        }
+    }
+
+    success = true;
+done:
+    Py_XDECREF(message_self);
+    Py_XDECREF(headers);
+    Py_XDECREF(map);
+    Py_XDECREF(method);
+    Py_XDECREF(path);
+
+    return success;
+}
+
+struct aws_http_message *aws_py_get_http_message(PyObject *http_message) {
+    struct aws_http_message *native = NULL;
+
+    PyObject *capsule = PyObject_GetAttrString(http_message, "_binding");
+    if (capsule) {
+        struct http_message_binding *message_binding = PyCapsule_GetPointer(capsule, s_capsule_name_http_message);
+        if (message_binding) {
+            /* Update contents of aws_http_message before returning it */
+            if (s_http_message_update_from_py(message_binding)) {
+                native = message_binding->native;
+                AWS_FATAL_ASSERT(native);
+            }
+        }
+        Py_DECREF(capsule);
+    }
+
+    return native;
+}
+
+PyObject *aws_py_http_request_new(PyObject *self, PyObject *args) {
+    (void)self;
+
+    PyObject *py_request;
+    PyObject *py_body_stream;
+    if (!PyArg_ParseTuple(args, "OO", &py_request, &py_body_stream)) {
+        return NULL;
+    }
+
+    AWS_FATAL_ASSERT(py_request != Py_None);
+
+    struct aws_allocator *alloc = aws_py_get_allocator();
+    struct http_message_binding *request = aws_mem_calloc(alloc, 1, sizeof(struct http_message_binding));
+    if (!request) {
+        return PyErr_AwsLastError();
+    }
+
+    /* From hereon, we need to clean up if errors occur.
+     * Fortunately, the capsule destructor will clean up anything stored inside the binding */
+
+    PyObject *capsule = PyCapsule_New(request, s_capsule_name_http_message, s_http_message_capsule_destructor);
+    if (!capsule) {
+        goto error;
+    }
+
+    /* Create native message now, but we don't set its contents till later */
+    request->native = aws_http_message_new_request(alloc);
+    if (!request->native) {
+        goto error;
+    }
+
+    request->self_proxy = PyWeakref_NewProxy(py_request, NULL);
+    if (!request->self_proxy) {
+        goto error;
+    }
+
+    /* If HttpMessage._body_stream set, create native aws_input_stream to read from it. */
+    if (py_body_stream != Py_None) {
+        request->native_body_stream = aws_input_stream_new_from_py(py_body_stream);
+        if (!request->native_body_stream) {
+            PyErr_SetAwsLastError();
+            goto error;
+        }
+
+        aws_http_message_set_body_stream(request->native, request->native_body_stream);
+    }
+
+    return capsule;
+
+error:
+    if (capsule) {
```

Comment:
This needs (because the proxy can fail):
```cpp
if(request->native) { 
    aws_http_message_destroy(request->native);
}
```

Reply:
nope! I have stumbled across a glorious python-only error handling technique:
Once you've created a capsule, it's destructor is guaranteed to run.
So I create a capsule around the binding struct as soon as possible. Then everything else that might need a teardown goes into the binding struct. If something goes wrong during initialization the capsule gets GC'd, the destructor runs, and everything gets cleaned up.

Reply:
Maybe a comment then? That was confusing to read.